src/coord/src/coord.rs

```diff
@@ -3074,6 +3079,67 @@ impl Coordinator {
         Ok(GlobalId::Transient(id))
     }
+
+    /// Return an error if the ids are from incompatible timelines. This should
+    /// be used to prevent users from doing things that are either meaningless
+    /// (joining data from timelines that have similar numbers with different
+    /// meanings like two separate debezium topics) or will never complete (joining
+    /// byo and realtime data).
+    fn validate_timeline(&self, mut ids: Vec<GlobalId>) -> Result<(), CoordError> {
+        let mut timelines = HashMap::new();
+
+        // Recurse through IDs to find all sources and tables, adding new ones to
+        // the set until we reach the bottom. Static views will end up with an empty
+        // timelines.
+        while let Some(id) = ids.pop() {
+            // Protect against possible infinite recursion. Not sure if it's possible, but
+            // a cheap prevention for the future.
+            if timelines.contains_key(&id) {
+                continue;
+            }
+            let entry = self.catalog.get_by_id(&id);
+            match entry.item() {
+                CatalogItem::Source(source) => {
+                    timelines.insert(id, source.connector.timeline());
+                }
+                CatalogItem::Index(index) => {
+                    ids.push(index.on);
+                }
+                CatalogItem::View(view) => {
+                    ids.extend(view.optimized_expr.as_ref().global_uses());
+                }
+                CatalogItem::Table(table) => {
+                    timelines.insert(id, table.timeline.clone());
+                }
+                _ => {}
+            }
+        }
+
+        let timelines: HashSet<Timeline> = timelines
+            .into_iter()
+            .map(|(_, timeline)| timeline)
+            .collect();
+
+        // If there's more than one timeline, we will not produce meaningful
+        // data to a user. Take, for example, some realtime source and a debezium
+        // consistency topic source. The realtime source uses something close to now
+        // for its timestamps. The debezium source starts at 1 and increments per
+        // transaction. We don't want to choose some timestamp that is valid for both
+        // of these because the debezium source will never get to the same value as the
+        // realtime source's "milliseconds since Unix epoch" value. And even if it did,
+        // it's not meaningful to join just because those two numbers happen to be the
+        // same now.
+        //
+        // Another example: assume two separate debezium consistency topics. Both
+        // start counting at 1 and thus have similarish numbers that probably overlap
+        // a lot. However it's still not meaningful to join those two at a specific
+        // transaction counter number because those counters are unrelated to the
+        // other.
+        if timelines.len() > 1 {
+            coord_bail!("Dataflow cannot use multiple timelines");
```

Comment:
We might in the future need to link to a more descriptive place in the site that tells folks what a timeline is / how to get past this error message (no need to do that before merging)

Reply:
Agreed. After I finish up linearizability I plan on writing docs for a week to describe all the stuff.